src/dma/mod.rs

```diff
@@ -0,0 +1,1171 @@
+        #[inline]
+        pub fn peripheral_burst(mut self, peripheral_burst: BurstMode) -> Self {
+            self.peripheral_burst = peripheral_burst;
+            self
+        }
+    }
+}
+
+/// DMA Transfer.
+pub struct Transfer<STREAM, CHANNEL, PERIPHERAL, DIRECTION, BUF>
+where
+    STREAM: Stream,
+    PERIPHERAL: PeriAddress,
+    BUF: WriteBuffer<Word = <PERIPHERAL as PeriAddress>::MemSize> + 'static,
+{
+    stream: STREAM,
+    _channel: PhantomData<CHANNEL>,
+    peripheral: PERIPHERAL,
+    _direction: PhantomData<DIRECTION>,
+    buf: Option<BUF>,
+    double_buf: Option<BUF>,
+    // Used when double buffering
+    transfer_length: u16,
+}
+
+impl<STREAM, CHANNEL, PERIPHERAL, DIR, BUF> Transfer<STREAM, CHANNEL, PERIPHERAL, DIR, BUF>
+where
+    STREAM: Stream,
+    CHANNEL: Channel,
+    DIR: Direction,
+    PERIPHERAL: PeriAddress,
+    BUF: WriteBuffer<Word = <PERIPHERAL as PeriAddress>::MemSize> + 'static,
+    (STREAM, CHANNEL, PERIPHERAL, DIR): Sealed,
+{
+    /// Applies all fields in DmaConfig.
+    fn apply_config(&mut self, config: config::DmaConfig) {
+        let msize = mem::align_of::<<PERIPHERAL as PeriAddress>::MemSize>() / 2;
+
+        self.stream.clear_interrupts();
+        self.stream.set_priority(config.priority);
+        // NOTE(unsafe) These values are correct because of the invariants of PeriAddress
+        unsafe {
+            self.stream.set_memory_size(msize as u8);
+            self.stream.set_peripheral_size(msize as u8);
+        }
+        self.stream.set_memory_increment(config.memory_increment);
+        self.stream
+            .set_peripheral_increment(config.peripheral_increment);
+        self.stream
+            .set_transfer_complete_interrupt_enable(config.transfer_complete_interrupt);
+        self.stream
+            .set_half_transfer_interrupt_enable(config.half_transfer_interrupt);
+        self.stream
+            .set_transfer_error_interrupt_enable(config.transfer_error_interrupt);
+        self.stream
+            .set_direct_mode_error_interrupt_enable(config.direct_mode_error_interrupt);
+        self.stream
+            .set_fifo_error_interrupt_enable(config.fifo_error_interrupt);
+        self.stream.set_double_buffer(config.double_buffer);
+        self.stream.set_fifo_threshold(config.fifo_threshold);
+        self.stream.set_fifo_enable(config.fifo_enable);
+        self.stream.set_memory_burst(config.memory_burst);
+        self.stream.set_peripheral_burst(config.peripheral_burst);
+    }
+
+    /// Configures the DMA stream to the correct channel for the peripheral, configures source and
+    /// destination and applies supplied configuration. In a memory to memory transfer,
+    /// the `double_buf` argument is the source of the data.
+    pub fn init(
+        mut stream: STREAM,
+        peripheral: PERIPHERAL,
+        mut memory: BUF,
+        mut double_buf: Option<BUF>,
+        config: config::DmaConfig,
+    ) -> Result<Self, DMAError<BUF>> {
```

Comment:
Hmm, not sure how people feel about panics, but I think you might be right on this one.

Reply:
Also not sure. I guess they can add fmt bloat if they don't get optimized away. But a `result.unwrap()` does too and that's probably harder for the compiler to optimize because of the additional level of indirection.

Maybe there is some crate-wide policy regarding Results vs panics that can help us here?